pkg/build/limit.go

```diff
@@ -0,0 +1,55 @@
+// Copyright 2018 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build
+
+import (
+	"context"
+
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"golang.org/x/sync/semaphore"
+)
+
+// Limiter composes with another Interface to limit the number of concurrent builds.
+type Limiter struct {
+	Builder   Interface
+	semaphore *semaphore.Weighted
+}
+
+// Limiter implements Interface
+var _ Interface = (*Recorder)(nil)
+
+// IsSupportedReference implements Interface
+func (l *Limiter) IsSupportedReference(ip string) bool {
+	return l.Builder.IsSupportedReference(ip)
+}
+
+// Build implements Interface
+func (l *Limiter) Build(ip string) (v1.Image, error) {
+	// TODO(jonjohnsonjr): Build should take a context.Context.
+	if err := l.semaphore.Acquire(context.TODO(), 1); err != nil {
```

Comment:
Is there any reason not to use [`sync.WaitGroup`](https://golang.org/pkg/sync/#WaitGroup)? I'm unfamiliar with the `semaphore` package, but it seems like it's main benefit is weightedness, which we're not taking advantage of here.

Reply:
WaitGroup is similar, but AFAICT there is no way to bound the maximum number of the counter, whereas NewWeighted takes a maximum weight.